cartographer/mapping/internal/2d/normal_estimation_2d.cc

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/mapping/internal/2d/normal_estimation_2d.h"
+
+namespace cartographer {
+namespace mapping {
+namespace {
+
+// Calulates the circular mean by transforming 'angles' to unit vectors and
+// calculating the arithmetic mean in Cartesian space.
+template <typename T>
+T CircularMean(const std::vector<T>& angles) {
+  T summed_y = 0;
+  T summed_x = 0;
+  for (const T angle : angles) {
+    summed_y += std::sin(angle);
+    summed_x += std::cos(angle);
+  }
+  return ceres::atan2(summed_y, summed_x);
```

Comment:
I think atan2 is undefined when x and y are 0.

Reply:
Atan2(0,0) requires the evaluation of 0/0, which it is not defined from a mathematical point of view.
If the atan2 implementation suppports IEEE floating-point arithmetics, it will return 0 or pi.